### ISSUE TITLE
Switch to tree-sitter

### ISSUE THREAD

**Issue by gcv** [gcv/julia-snail#149]
By switching to tree-sitter, built into Emacs as of version 29, all the questions Snail asks of CSTParser can be answered without leaving Elisp. This will be a significant benefit, not only by eliminating a dependency on CSTParser and its potentially unstable API, but also because Julia startup time will improve.

In general, it'll be great to ditch all Julia-side dependencies for core Snail functionality and leave them to extensions.

**Comment by MasonProtter** [comment]
By the way, JuliaSyntax.jl is now bundled inside of modern versions of Julia, so you *could* also just use JuliaSyntax.jl instead of TreeSitter, but if there's other benefits to TreeSitter then I guess we should go with that.

**Comment by linwaytin** [comment]
I also feel `JuliaSyntax.jl` should be a better option than TreeSitter. The progress of [tree-sitter-julia](https://github.com/tree-sitter/tree-sitter-julia) is slow, and I do not think it is supported officially from the Julia community.

**Comment by gcv** [comment]
At this point, I’m also leaning in the direction of JuliaSyntax. I’m a little concerned about yanking the rug out from any users who need Julia versions <1.10, but it should be okay, as long as the docs point out that older versions of Snail can be installed using straight.el or Elpaca.

**Comment by linwaytin** [comment]
Also I think `JuliaSyntax` can be manually installed for version < 1.10, so it should not be a big problem to use this parser.

**Comment by MasonProtter** [comment]
Yes, JuliaSyntax is compatible all the way back to julia v1.0 and is on the general registry.

**Comment by gcv** [comment]
@MasonProtter, @linwaytin, @danielmatz, @karthink, @dahtah, @alecStewart1, @norci, @orialb: I now have a branch with CSTParser replaced with JuliaSyntax: https://github.com/gcv/julia-snail/tree/juliasyntax

Unless someone persuades me otherwise, I will not bother supporting Julia <1.10 at this point. I like that Snail doesn't have anything in its Project.toml now.

It could use some testing before I merge into the main branch and break everything for everyone on MELPA. 😀

**Comment by linwaytin** [comment]
Thanks for working on this. 

I tried it but I keep getting errors when typing. A `*julia* error` buffer pops up showing
```
MethodError: no method matching pathat(::Nothing, ::Int64)

Closest candidates are:
  pathat(!Matched::Base.JuliaSyntax.SyntaxNode, ::Any, !Matched::Any)
   @ Main.JuliaSnail ~/new-emacs.d/elpa/julia-snail/JuliaSnail.jl:533
  pathat(!Matched::Base.JuliaSyntax.SyntaxNode, ::Any)
   @ Main.JuliaSnail ~/new-emacs.d/elpa/julia-snail/JuliaSnail.jl:533


moduleat(encodedbuf::String, byteloc::Int64) at JuliaSnail.jl:634
top-level scope at none:1
eval at boot.jl:385 [inlined]
eval_in_module(fully_qualified_module_name::Vector{Symbol}, expr::Expr) at JuliaSnail.jl:245
(::Main.JuliaSnail.var"#30#36"{Sockets.TCPSocket})() at JuliaSnail.jl:1014
```
Also in the `*julia*` buffer there is `Base.JuliaSyntax.ParseError`.

It seems the package tries to parse the code in the middle when I'm typing, but since the code is incomplete `JuliaSyntax` cannot parse it successfully.

**Comment by gcv** [comment]
@linwaytin: Thank you for testing! @MasonProtter sent in a fix, and I fixed another bug. Please try the branch again.

There's also a public PR for this branch now: #176

**Comment by linwaytin** [comment]
Thanks for such a quick fix. I gave a simple try. The error I met is gone. Awesome!

Typing with `snail` activated seems a little delayed. From my memory, it seems slower than the version with the old parser.

**Comment by gcv** [comment]
It's certainly possible that JuliaSyntax is slower that CSTParser. I don't know, and this is probably worth benchmarking. I'm curious, however, why your environment uses the parser continuously while you type. Do you have have some kind of autocomplete running at all times? I'd like to reproduce the slow typing problem... hitting the parser at each keystroke is highly inefficient, and maybe there's a way to cache the results or debounce it.

**Comment by linwaytin** [comment]
Yes, I'm using `company`. Maybe it is problem? Should I turn it off when using `snail`?

**Comment by gcv** [comment]
You shouldn't have to turn off a coding aid you find useful. I just pushed a change that caches some intermediate parser results that are used by various completion frameworks (4774334a0a). It's a little rough, but please try it and let me know if it made things better.

**Comment by linwaytin** [comment]
It's quite smooth now! Thanks for the quick fix!

**Comment by linwaytin** [comment]
A side issue: The completion in `org-mode` is broken in this branch. When I type in a src block in `org-mode`, I get 
```
UndefVarError: `CST` not defined
```

**Comment by gcv** [comment]
Thanks, that should be fixed in https://github.com/gcv/julia-snail/commit/1bfa18bc300be54efd83a8f1e3e51d725141067c.

**Comment by gcv** [comment]
I’ve been using this branch for a few weeks, and it seems to generally work well for me. I would merge it, but I’ve been thinking that removing CSTParser compatibility wholesale is too aggressive. I think that many, many Snail users still rely on older versions of Julia: cluster environments, sensitive research, and other excellent reasons to stick to something older stable.

If Emacs packaging supported proper version pinning, I’d just suggest these users install a CSTParser-compatible version of Snail. Unfortunately, given the limitations of package.el and ELPA-style package repositories, that’s not an option unless I release a different package under a different name, which I dislike for many reasons. (And I’m not telling everyone to go and switch from package.el to Elpaca, either.)

So I think Snail will need to retain backwards compatibility with old versions of Julia for a long time. At the same time, I don’t want to give up on JuliaSyntax — it’s a better experience for me personally, and it fixes packaging on read-only systems like Nix. So I think I have to pull a Microsoft: make the code dirtier and harder to maintain, but aggressively backwards-compatible with older environments. This won’t happen immediately, but I’ll put back CSTParser integration, and make sure it’s only enabled for older versions of Snail.

**Comment by linwaytin** [comment]
How about keeping two branches? Currently, I use `package-vc`, which is built-in in Emacs 30, to check out the `juliasyntax` branch. It's very simple. Actually, I think it can also pin to a particular version. 

`JuliaSyntax.jl` has been included in Julia 1.10, and it can be installed in earlier versions too, so I do not think the backward compatibility is a big issue. 

Just my two cents.

**Comment by MasonProtter** [comment]
I don't know if I'm just being annoying by saying this over and over, or if it just wasn't noticed the previous times I said it, but JuliaSyntax is compatible and runnable on all versions of julia going back to version 1.0. I don't understand why you want to ti maintain two different parsers when you can just use JuliaSyntax on all versions. 

You can just check at startup if the julia version is newer than 1.9, and if it is, then use `Base.JuliaSyntax`, and if it's older, then you use your old trick of inserting a `Project.toml` or whatever with JuliaSyntax in it.

**Comment by gcv** [comment]
I tried JuliaSyntax on older versions of Julia early on, and something didn’t work. I don’t remember what. It’s worth trying again, though, now that I have more experience with it.
